keyboards/dp60/keymaps/indicator/indicator.c

```diff
@@ -0,0 +1,253 @@
+/*
+ * light weight WS2812 lib V2.0b
+ *
+ * Controls WS2811/WS2812/WS2812B RGB-LEDs
+ * Author: Tim (cpldcpu@gmail.com)
+ *
+ * Jan 18th, 2014  v2.0b Initial Version
+ * Nov 29th, 2015  v2.3  Added SK6812RGBW support
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 2 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+/**
+ * indicator.c
+ */
+
+#include "dp60.h"
+
+#include "rgblight_list.h"
+#include "rgblight.h"
+#include "ws2812.h"
+#include <avr/interrupt.h>
+#include <avr/io.h>
+#include <util/delay.h>
+
+/*
+  This routine writes an array of bytes with RGB values to the Dataout pin
+  using the fast 800kHz clockless WS2811/2812 protocol.
+*/
+
+// Timing in ns
+#define w_zeropulse 350
+#define w_onepulse 900
+#define w_totalperiod 1250
+
+// Fixed cycles used by the inner loop
+#define w_fixedlow 2
+#define w_fixedhigh 4
+#define w_fixedtotal 8
+
+// Insert NOPs to match the timing, if possible
+#define w_zerocycles (((F_CPU / 1000) * w_zeropulse) / 1000000)
+#define w_onecycles (((F_CPU / 1000) * w_onepulse + 500000) / 1000000)
+#define w_totalcycles (((F_CPU / 1000) * w_totalperiod + 500000) / 1000000)
+
+// w1 - nops between rising edge and falling edge - low
+#define w1 (w_zerocycles - w_fixedlow)
+// w2   nops between fe low and fe high
+#define w2 (w_onecycles - w_fixedhigh - w1)
+// w3   nops to complete loop
+#define w3 (w_totalcycles - w_fixedtotal - w1 - w2)
+
+#if w1 > 0
+#    define w1_nops w1
+#else
+#    define w1_nops 0
+#endif
+
+// The only critical timing parameter is the minimum pulse length of the "0"
+// Warn or throw error if this timing can not be met with current F_CPU settings.
+#define w_lowtime ((w1_nops + w_fixedlow) * 1000000) / (F_CPU / 1000)
+#if w_lowtime > 550
+#    error "Light_ws2812: Sorry, the clock speed is too low. Did you set F_CPU correctly?"
+#elif w_lowtime > 450
+#    warning "Light_ws2812: The timing is critical and may only work on WS2812B, not on WS2812(S)."
+#    warning "Please consider a higher clockspeed, if possible"
+#endif
+
+#if w2 > 0
+#    define w2_nops w2
+#else
+#    define w2_nops 0
+#endif
+
+#if w3 > 0
+#    define w3_nops w3
+#else
+#    define w3_nops 0
+#endif
+
+#define w_nop1 "nop      \n\t"
+#define w_nop2 "rjmp .+0 \n\t"
+#define w_nop4 w_nop2 w_nop2
+#define w_nop8 w_nop4 w_nop4
+#define w_nop16 w_nop8 w_nop8
+void inline indicator_sendarray_mask_ind(uint8_t *data, uint16_t datlen, uint8_t maskhi) {
```

Comment:
Not a fan of this duplication of core code. Im guessing the only requirement is to be able to use 2 different RGB pins, and if so, theres ways to avoid having to re-implement the ws2812 driver.

Reply:
The current avr ws2812 driver embedded the PIN definition on the assembly codes. I'm not familiar with assembly and have no idea how to extract the PIN definition from the codes. Any help is great.

Reply:
IIRC, we've had this talk on discord (not you specifically), but there is no way to do this without ... basically re-implementing it, as the data pin HAS to be defined at compile time, and can't be changed in runtime. 

This is also why #9412 was opened.

Reply:
I think was the main thing holding this back. 

Another way to do this, maybe would be to undefine the rgbled pin, define the new version, and then include ws2812.c.  Which, is super hacky, but wouldn't duplicate core code.